_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,193 @@
+
+### Reusability
+
+**When a component has a single, narrow responsibility, it's easier to reuse it.** `FraudDetection` class from the previous section is an example of this: we could reuse it in online processing and batch processing components. To do this in the *artifact space*, we could pull it into a shared library. Another direction is to move fraud detection into a separate microservice, which we can think about as reusability in the *runtime space*. The `FraudDetection` class within our application will then turn from having business logic responsibility to do external integration with the new service.
+
+**Most methods with a narrow responsibility shouldn't have side effects and shouldn't depend on the state of the class**, which enables sharing and calling them from any place. In other words, Single Responsibility Principle nudges us toward [functional programming](https://en.wikipedia.org/wiki/Functional_programming) style.
+
+***
+Pro tip: thinking about responsibilities helps to notice [unrelated subproblems](https://learning.oreilly.com/library/view/the-art-of/9781449318482/ch10.html) hiding in our methods and classes. When we extract them, we can then see opportunities to reuse them in other places. Moving unrelated subproblems out of the way keeps a component at the [single level or abstraction](http://principles-wiki.net/principles:single_level_of_abstraction), which makes easier to understand the logic of the component.
+***
+
+### Testability
+
+**It's easier to test methods and classes with focused, independent concerns.** This is what [Humble Object](http://xunitpatterns.com/Humble%20Object.html) pattern is all about. Let's continue playing with `TransactionProcessor`:
+```
+class TransactionProcessor {
+    
+  void processTransaction(Transaction t) {
+    boolean isFraud;
+    // logic detecting that the transaction is fraud, many
+    // line of code omitted
+    ...
+    
+    if (isFraud) {
+      logger.log("Detected fraud transaction {}", t);
+      alertingService.sendAlert(new FraudTransactionAlert(t));
+    }
+  }
+}
+```
+In this variant, there is no separate `isFraud()` method. `processTransaction()` conflates the fraud detection and the reporting logic. Then, to test the fraud detection, we either need to mock the `alertingService`, which pollutes the test code with boilerplate, or to intercept and check the logging output, which is also cumbersome, and may also hinder our ability to execute tests in parallel.
```

Comment:
```suggestion
In this variant, there is no separate `isFraud()` method. `processTransaction()` combines the fraud detection and the reporting logic. To test the fraud detection logic, we don't need the alerting logic. So, we either need to mock the `alertingService`, which pollutes the test code with boilerplate, or we intercept and check the logging output, which is also cumbersome, and may even hinder our ability to execute tests in parallel.
```

Reply:
Why conflates -> combines? I want to have a negative connotation.

Reply:
Also rephrased this part separately, please review again.